python/hail/utils/hadoop_utils.py

```diff
@@ -4,134 +4,74 @@
+     - ``'w'`` -- Writable text file (:class:`io.TextIOWrapper`).
+     - ``'rb'`` -- Readable binary file (:class:`io.BufferedReader`).
+     - ``'wb'`` -- Writable binary file (:class:`io.BufferedWriter`).
 
-        These file handles are slower than standard Python file handles.
-        If you are reading a file larger than ~50M, it will be faster to
-        use :meth:`~hail.hadoop_copy` to copy the file locally, then read it
-        with standard Python I/O tools.
-
-    Parameters
-    ----------
-    path: :obj:`str`
-        Source file URI.
-    buffer_size: :obj:`int`
-        Size of internal buffer.
-
-    Returns
-    -------
-    `io.BufferedReader`
-        Binary file reader,
-        `io.BufferedReader <https://docs.python.org/3/library/io.html#io.BufferedReader>`__.
-    """
-    return io.BufferedReader(HadoopReader(path, buffer_size), buffer_size=buffer_size)
-
-
-@typecheck(path=str,
-           buffer_size=int)
-def hadoop_write(path, buffer_size=8192):
-    """Open a writable file through the Hadoop filesystem API.
-    Supports distributed file systems like hdfs, gs, and s3.
-
-    Examples
-    --------
-
-    .. doctest::
-        :options: +SKIP
-
-        >>> with hadoop_write('gs://my-bucket/notes.txt') as f:
-        ...     f.write('result1: %s\\n' % result1)
-        ...     f.write('result2: %s\\n' % result2)
-
-    Notes
-    -----
     The provided destination file path must be a URI (uniform resource identifier).
 
     .. caution::
 
         These file handles are slower than standard Python file handles. If you
         are writing a large file (larger than ~50M), it will be faster to write
-        to a local file using standard Python I/O and use :meth:`~hail.hadoop_copy`
+        to a local file using standard Python I/O and use :func:`.hadoop_copy`
         to move your file to a distributed file system.
 
     Parameters
     ----------
-    path: :obj:`str`
-        Destination file URI.
+    path : :obj:`str`
+        Path to file.
+    mode : :obj:`str`
+        File access mode.
+    buffer_size : :obj:`int`
+        Buffer size, in bytes.
 
     Returns
     -------
-    :class:`io.BufferedWriter`
-        File writer object,
-        `io.BufferedWriter <https://docs.python.org/3/library/io.html#io.BufferedWriter>`__.
+        Readable or writable file handle.
     """
-    return io.TextIOWrapper(io.BufferedWriter(HadoopWriter(path), buffer_size=buffer_size), encoding='iso-8859-1')
+    supported_modes = ('r', 'w', 'rb', 'wb')
+    if not mode in supported_modes:
+        raise ValueError(f"unsupported mode {repr(mode)}. Supported modes are {supported_modes}")
+
+
+    if 'r' in mode:
+        handle = io.BufferedReader(HadoopReader(path, buffer_size), buffer_size=buffer_size)
+    else:
+        handle = io.BufferedWriter(HadoopWriter(path), buffer_size=buffer_size)
+
+    if 'b' in mode:
+        return handle
+    else:
+        return io.TextIOWrapper(handle, encoding='iso-8859-1')
```

Comment:
I feel like `encoding` should be an extra argument to `hadoop_open`? iso-8859-1 seems like an odd default to me, but maybe it's sensible in the bioinformatics world? I'd think we'd just do UTF-8.

Reply:
the encoding must agree with the JVM string encoding. I've tried to parameterize, but anything other than iso-8859-1 breaks. It's probably possible to go and add the parameter on both sides of py4j, however (though I don't want to do that here).

Reply:
the native, in-memory JVM string encoding should be UTF-16, does that encoding fail?

Reply:
I'll test, but I think so.